R/filters.R

```diff
@@ -1,70 +1,115 @@
+#' Possible values:
+#' - exact: match only the exact specified string;
+#' - startswith: match all strings starting with the specified substring;
+#' - substring: match all strings containing the specified substring.
+#' Default value: 'exact'.
 #'
 #' @examples
 #' data(immdata)
 #'
 #' # Select samples with status "MS"
-#' repFilter(immdata, "by.meta", list(Status = "MS"))
+#' repFilter(immdata, "by.meta", list(Status = include("MS")))
+#'
+#' # Select samples without status "MS"
+#' repFilter(immdata, "by.meta", list(Status = exclude("MS")))
+#'
+#' # Select samples from lanes "A" and "B" with age > 15
+#' repFilter(immdata, "by.meta", list(Lane = include("A", "B"), Age = morethan(15)))
+#'
+#' # Select samples that are not from lanes "A" and "B"
+#' repFilter(immdata, "by.meta", list(Lane = exclude("A", "B")))
 #'
-#' # Select samples from lanes "A" and "B" with age >= 15
-#' repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=15"))
+#' # Select samples with 1000-5000 clonotypes
+#' repFilter(immdata, "by.repertoire", list(n_clonotypes = interval(1000, 5000)))
+#'
+#' # Select clonotypes in all samples with alpha chains
+#' repFilter(immdata, "by.clonotype",
+#'   list(V.name = include("AV"), J.name = include("AJ")),
+#'   match = "substring"
+#' )
 #' @export repFilter
-repFilter <- function(.data, .method, .query, ...) {
+repFilter <- function(.data, .method = "by.clonotype",
+                      .query = list(CDR3.aa = exclude("partial", "out_of_frame")),
+                      .match = "exact") {
   if (!is.list(.data)) {
     stop(paste0(
       "Input data is not a list; ",
-      "please pass Immunarch object with data and metadata as input."
+      "please pass Immunarch dataset object as input."
     ))
   }
-  if (!("meta" %in% names(.data))) {
+
+  if (!(.match %in% c("exact", "startswith", "substring"))) {
     stop(paste0(
-      "Input data doesn't contain meta; ",
-      "please pass Immunarch object with data and metadata as input."
+      "Unknown matching method \"", .match,
+      "\"! Supported matching methods are: ",
+      "\"exact\", \"startswith\", \"substring\"."
     ))
   }
 
   switch(tolower(.method),
-    by.meta = filter_by_meta(.data, .query),
-    stop("You entered the wrong method! Supported methods are: \"by.meta\".")
+    by.meta = filter_by_meta(.data, .query, .match),
+    by.repertoire = filter_by_repertoire(.data, .query),
+    by.rep = filter_by_repertoire(.data, .query),
+    by.clonotype = filter_by_clonotype(.data, .query, .match),
+    by.cl = filter_by_clonotype(.data, .query, .match),
+    stop(paste0(
+      "You entered wrong method \"", .method, "\"! Supported methods are: ",
+      "\"by.meta\", \"by.repertoire\", \"by.clonotype\"."
+    ))
   )
 }
 
-filter_by_meta <- function(.data, .query) {
-  names <- names(.query)
+filter_by_meta <- function(.data, .query, .match) {
+  if (!("meta" %in% names(.data))) {
```

Comment:
unfortunately, there is no convention that meta table should have name 'meta', user can rename it, only column names in data are fixed; but you can check is in this object list and data.frame and data.frame will be meta and list will be data;
Example of usage: `repFilter(list(list_with_repertoirs, table_with_custom_meta), 'by.meta' ,list(Sex=include('M')))`

Reply:
Also, if names are wrong you can warn that after repFilter user will receive object with data and meta names instead of custom.